### PR TITLE
Azure Cosmos DB clustering/persistence/reminder provider

### PR DESCRIPTION
**Pull request by pentp**
This is a new implementation using Cosmos DB using its native API.
It should be more reliable, faster, less costly and the state storage max size is more than doubled.
The grain state, reminders and clustering providers have been in production use for almost a year already, only the grain directory provider hasn't seen production use.

### CODE REVIEW

File: src/Azure/Orleans.AzureCosmos/AzureCosmosGrainStorage.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                if (res.StatusCode == HttpStatusCode.NotFound)
+                {
+                    if (logger.IsEnabled(LogLevel.Trace)) logger.Trace("NotFound reading: PartitionKey={0} RowKey={1} from Container={2}", pk, grainType, options.ContainerName);
+                    return;
+                }
+
+                var eTag = res.EnsureSuccessStatusCode().Headers.ETag;
+                if (logger.IsEnabled(LogLevel.Trace)) logger.Trace("Read: PartitionKey={0} RowKey={1} from Container={2} with ETag={3}", pk, grainType, options.ContainerName, eTag);
+
+                var record = (GrainStateRecord)serializer.Deserialize(new StreamReader(res.Content), typeof(GrainStateRecord<>).MakeGenericType(grainState.Type));
```
Comment (ReubenBond) [inline src/Azure/Orleans.AzureCosmos/AzureCosmosGrainStorage.cs:68]: How cheap is this MakeGenericType call, does it make sense to cache the results?

**Comment by ReubenBond** [comment]
How does this compare with the [CosmosDB providers in OrleansContrib](https://github.com/OrleansContrib/Orleans.CosmosDB)? Should the two be converged into a single set of packages in this repo?

**Comment by KSemenenko** [comment]
In my experience retry is necessary. From time to time something goes wrong or the query limits are triggered

**Comment by ReubenBond** [comment]
I'll close this due to inactivity, but please feel free to re-open, especially if there are responses to the above comments.

**Pull request closed without merging**